go/pkg/cas/upload.go

```diff
@@ -21,8 +23,19 @@ import (
 	"github.com/bazelbuild/remote-apis-sdks/go/pkg/digest"
 	"github.com/bazelbuild/remote-apis-sdks/go/pkg/retry"
 	repb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
 )
 
+// zstdEncoders is a pool of ZStd encoders.
+// Clients of this pool must call Close() on the encoder after using the
+// encoder.
+var zstdEncoders = sync.Pool{
+	New: func() interface{} {
+		enc, _ := zstd.NewWriter(nil)
```

Comment:
could you elaborate what do we gain from this? 

In my experiments, most of the time is spent on uploading large files which are uploaded sequentially (most small files are already present). If we use concurrency=1, then we would be using only one CPU per `cas.Client` most of the time.

Reply:
`zstd.NewWriter()` spawns goroutines internally, IIRC it defauts to the number of virtual CPU cores so `Encoder.EncodeAll()` can process that many simultaneous encodes. But if you're using a `sync.Pool` you're probably only ever doing one encode per `Encoder`. If you create these with `WithEncoderConcurrency(1)` then it creates fewer goroutines.

Reply:
@nodirg I'm not sure of the specifics of your use case, but unless you have a large tail end of files < ncores, you'll have enough parallelism controlled by `semByteStreamWrite`.

Few tail ends would be problematic with one concurrency though.

Reply:
see `muLargeFiles`: we upload at most one large file at a time (because it results in higher readers from `pd-standard` disks), and this is where most of the upload time goes (in my case, 22min to upload ~50GB, where vasy majority of bytes reside in few large files). So we have 8 cores (e2-standard-8 machine) with one upload at a time.

to be super clear: `muLargeFiles` limits concurrency much more than `semByteStramWrite` - we can ignore the latter in this case.